R/process_source.R

```diff
@@ -1,131 +1,71 @@
+  text <- readLines(file)
+
+  assignment_ops <- parse_data[
+    parse_data$token %in% c("LEFT_ASSIGN", "EQ_ASSIGN", "RIGHT_ASSIGN"),
+  ]
+  # Assignments are the parent expressions of assignment operators
+  assignments <- parse_data[parse_data$id %in% assignment_ops$parent, ]
+  # Top-level expressions have a parent attribute of 0
+  object_definitions <- assignments[assignments$parent == 0, ]
+
+  staticexports <- vector("list", nrow(object_definitions))
+
+  for (i in seq_len(nrow(object_definitions))) {
+    object_definition <- object_definitions[i, ]
+
+    names(staticexports)[i] <- extract_object_name(
+      parse_data, object_definition, assignment_ops
+    )
+
+    # A leading comment's parent attribute is 0 - the next expression's id
+    leading_comments <- parse_data[parse_data$parent == -object_definition$id, ]
+
+    # Exclude leading comments if there are any empty lines between the comment
+    # and the object definition
+    leading_comments <- leading_comments[
+      object_definition$line1 - leading_comments$line1 <=
+        rev(seq_len(nrow(leading_comments))),
+    ]
+
+    staticexport_lines <- seq(
+      min(leading_comments$line1, object_definition$line1),
+      max(leading_comments$line2, object_definition$line2)
+    )
+    staticexport_text <- text[staticexport_lines]
+    staticexports[[i]] <- staticexport_text
   }
 
-  chunks
-}
-
-
-# A wrapper object for vectors, which makes it easy to efficiently append items
-# to it.
-vec_accumulator <- function(init = logical(0)) {
-  s <- init
-  list(
-    add = function(x) {
-      s[length(s) + seq_along(x)] <<- x
-    },
-    get = function() {
-      s
-    },
-    reset = function() {
-      s <<- logical(0)
-    }
-  )
+  staticexports
 }
 
+extract_object_name <- function(parse_data, object_definition, assignment_ops) {
+  assignment <- assignment_ops[assignment_ops$parent == object_definition$id, ]
 
-is_parseable <- function(txt) {
-  parseable <- NULL
-  tryCatch({
-    parse(text = txt)
-    return(TRUE)
-  },
-    error = function(e) {}
-  )
-  return(FALSE)
-}
+  object_name_expr <- parse_data[
+    parse_data$parent == object_definition$id &
+      if (assignment$token %in% c("LEFT_ASSIGN", "EQ_ASSIGN")) {
+        parse_data$line2 <= assignment$line1 & parse_data$col2 < assignment$col1
+      } else {
+        parse_data$line1 >= assignment$line2 & parse_data$col1 > assignment$col2
+      },
+  ]
 
-# Given text like "foo <- function() {", return "foo". Also handles the case
-# where the name is surounded with backticks.
-extract_object_name <- function(txt) {
-  result <- sub("^(\\S+).*", "\\1", txt)
+  result <- utils::getParseText(parse_data, object_name_expr$id)
 
   # Remove backticks, so that strings like "`%||%`" are converted to "%||%"
   if (grepl("^`\\S+`$", result)) {
-    result <- sub("^`(\\S+)`$", "\\1", result)
+    result <- sub("^`(\\S+)`$", "\\1", result, perl = TRUE)
```

Comment:
Is the `perl=TRUE` change necessary?

Reply:
My motivation was that `perl = TRUE` gives better performance:

``` r
result <- c("x", "`%%`")
bench::mark(
  "perl = FALSE" = sub("^`(\\S+)`$", "\\1", result),
  "perl = TRUE" = sub("^`(\\S+)`$", "\\1", result, perl = TRUE)
)
#> # A tibble: 2 × 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 perl = FALSE   20.9µs   22.2µs    40286.        0B        0
#> 2 perl = TRUE    19.4µs   20.3µs    46868.        0B        0
```

<sup>Created on 2022-10-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

However, according to [this Stack Overflow post](https://stackoverflow.com/questions/47240375/regular-expressions-in-base-r-perl-true-vs-the-default-pcre-vs-tre#answer-47251004), it looks like the opposite might be true on Windows.

Given that the performance benefits aren't universal, I reverted in [0085e19](https://github.com/wch/staticimports/pull/21/commits/0085e19590b2d99cbc04cf0d32a0449dfae215e1).